diff --git a/tools/download_wgpu_native.py b/tools/download_wgpu_native.py
index dfc17c0..6a18acf 100644
--- a/tools/download_wgpu_native.py
+++ b/tools/download_wgpu_native.py
@@ -6,8 +6,6 @@ import tempfile
 import platform
 from zipfile import ZipFile
 
-import requests
-
 
 DEFAULT_UPSTREAM = "gfx-rs/wgpu-native"
 
@@ -49,6 +47,8 @@ def write_current_version(version, commit_sha):
 
 
 def download_file(url, filename):
+    import requests
+
     resp = requests.get(url, stream=True)
     with open(filename, mode="wb") as fh:
         for chunk in resp.iter_content(chunk_size=1024 * 128):
