diff --git a/pyproject.toml b/pyproject.toml
index 447d64c..10c89de 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -70,11 +70,6 @@ force-include = { "tools" = "tools" }
 packages = ["wgpu"]
 artifacts = ["*.so", "*.dll", "*.dylib"]
 
-# We use a hatch build hook to install the correct wgpu-native lib right before
-# the wheel is build, and to allow cross-platform builds. See the tools dir.
-[tool.hatch.build.targets.wheel.hooks.custom]
-path = "tools/hatch_build.py"
-
 [tool.pytest.ini_options]
 asyncio_default_fixture_loop_scope = "function"
