diff --git a/wgpu/backends/wgpu_native/_api.py b/wgpu/backends/wgpu_native/_api.py
index 320d122..5bdfbf4 100644
--- a/wgpu/backends/wgpu_native/_api.py
+++ b/wgpu/backends/wgpu_native/_api.py
@@ -2058,8 +2058,10 @@ class GPUDevice(classes.GPUDevice, GPUObjectBase):
         primitive_extras = {}
         if isinstance(primitive, dict):
             # in case of extras, the struct isn't used but a dict... so we check for it here.
-            primitive_extras["polygon_mode"] = primitive.pop("polygon_mode", "Fill")
-            primitive_extras["conservative"] = primitive.pop("conservative", False)
+            if (polygon_mode := primitive.pop("polygon_mode", None)) is not None:
+                primitive_extras["polygon_mode"] = polygon_mode
+            if (conservative := primitive.pop("conservative", None)) is not None:
+                primitive_extras["conservative"] = conservative
         check_struct("VertexState", vertex)
         check_struct("DepthStencilState", depth_stencil)
         check_struct("MultisampleState", multisample)
@@ -2086,19 +2088,22 @@ class GPUDevice(classes.GPUDevice, GPUObjectBase):
         )
 
         # explanations for extras: https://docs.rs/wgpu/latest/wgpu/struct.PrimitiveState.html
-        polygon_mode = enum_str2int["PolygonMode"].get(
-            primitive_extras.get("polygon_mode"), enum_str2int["PolygonMode"]["Fill"]
-        )
+        if len(primitive_extras) == 0:
+            next_in_chain = ffi.NULL
+        else:
+            polygon_mode = enum_str2int["PolygonMode"].get(
+                primitive_extras.get("polygon_mode"), enum_str2int["PolygonMode"]["Fill"]
+            )
 
-        # H: chain: WGPUChainedStruct, polygonMode: WGPUPolygonMode, conservative: WGPUBool/int
-        c_primitive_state_extras = new_struct_p(
-            "WGPUPrimitiveStateExtras *",
-            # not used: chain
-            polygonMode=polygon_mode,
-            conservative=primitive_extras.get("conservative", False),
-        )
-        c_primitive_state_extras.chain.sType = lib.WGPUSType_PrimitiveStateExtras
-        next_in_chain = ffi.cast("WGPUChainedStruct *", c_primitive_state_extras)
+            # H: chain: WGPUChainedStruct, polygonMode: WGPUPolygonMode, conservative: WGPUBool/int
+            c_primitive_state_extras = new_struct_p(
+                "WGPUPrimitiveStateExtras *",
+                # not used: chain
+                polygonMode=polygon_mode,
+                conservative=primitive_extras.get("conservative", False),
+            )
+            c_primitive_state_extras.chain.sType = lib.WGPUSType_PrimitiveStateExtras
+            next_in_chain = ffi.cast("WGPUChainedStruct const *", c_primitive_state_extras)
 
         # H: nextInChain: WGPUChainedStruct *, topology: WGPUPrimitiveTopology, stripIndexFormat: WGPUIndexFormat, frontFace: WGPUFrontFace, cullMode: WGPUCullMode, unclippedDepth: WGPUBool/int
         c_primitive_state = new_struct(
